### PR TITLE
Android Contacts Module: Significant performance increase

### PR DESCRIPTION
**Pull request by tiny-dancer**
cleaned up pr from here: https://github.com/expo/expo/pull/1188
relevant tests: https://github.com/expo/test-suite/pull/4

**significant performance improvement**

from [test-suite](https://github.com/expo/test-suite/pull/4)
- querying 20 contacts
  - SDK 25: 2000ms
  - PR: 150-250ms
- querying 500 contacts
  - SDK 25: Jasmine timeout over 5000ms
  - PR: 150-250ms

from a calling sample app
- querying 100 contacts
  - SDK 25: 7,500ms
  - PR: 2,500ms
- querying 100 contacts with email + phone
  - SDK 25: 9,500ms
  - PR:  3,100ms
- querying 500 contacts with email + phone
  - SDK 25: 26,000ms
  - PR: 6,500ms

based off of react-native-contact's implementation and centers around pulling necessary data from a single cursor query

this pull request:
image

sdk 25:
image

side notes
- resolves an "appears to be" defect around actual birthdate could be overwritten with another event date (no validation event date being set matches birthday label)
```
if (birthday) {   
    contact.putMap("birthday", details); 
}
```
- Primary field now returns if true or false, previously only returned if true

if IOS follows similar performance baselines to those in this PR,  I would propose in a future release looking into returning more fields by defaul. ie. phone, email and thumbnail path. 

thanks for looking!

### CODE REVIEW

**Comment by tiny-dancer** [comment]
@aalices here's updated pr from previous conversations, perf numbers are looking good! :)

**Comment by tiny-dancer** [comment]
@aalices gentle poke :) hoping to eventually layer the api changes on this foundation for android side

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 1]
```diff
@@ -42,6 +42,53 @@
 public class ContactsModule extends ReactContextBaseJavaModule {
   private static final String TAG = ContactsModule.class.getSimpleName();
 
+  private static final String[] PROJECTION = new String[]{
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: Can you make sure we really need the projections? From my quick check: this also works when passing null to ContentResolver's query but maybe I'm missing something?

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 2]
```diff
@@ -61,116 +108,127 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
               break;
-            } else {
-              contacts.pushMap(contact);
-              currentIndex++;
             }
+            contactsArray.pushMap(contact.toMap(fieldsSet));
           }
         }
 
         int total = cursor.getCount();
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: This return a wrong value - for example is 15 when my device has 6 contacts.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 3]
```diff
@@ -187,408 +245,303 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
-          }
-        }
+    while (cursor.moveToNext()) {
+      int columnIndex = cursor.getColumnIndex(ContactsContract.Data.CONTACT_ID);
+      String contactId;
+      if (columnIndex != -1) {
+        contactId = cursor.getString(columnIndex);
+      } else {
+        //todo - double check this, it may not be necessary any more
+        contactId = String.valueOf(-1);//no contact id for 'ME' user
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: have you done some research about this? What actually is a 'me' user?

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 4]
```diff
@@ -187,408 +245,303 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
-          }
+  private boolean isMissingPermissions() {
+    return Build.VERSION.SDK_INT >= 23 &&
+        ContextCompat.checkSelfPermission(
+            getReactApplicationContext(),
+            Manifest.permission.READ_CONTACTS) != PackageManager.PERMISSION_GRANTED;
+  }
 
-          String label;
+  private static class Contact {
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: How about making a package inside `api` and putting all the java classes as separate files? This one is getting huge and hard to read.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 5]
```diff
@@ -187,408 +245,303 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
+        contact.putString("phoneticLastName", phoneticLastName);
+      }
+      if (!TextUtils.isEmpty(phoneticMiddleName)) {
+        contact.putString("phoneticMiddleName", phoneticMiddleName);
+      }
+      contact.putString("company", company);
+      contact.putString("jobTitle", jobTitle);
+      contact.putString("department", department);
+      contact.putBoolean("imageAvailable", this.hasPhoto);
+      //contact.putString("thumbnailPath", photoUri == null ? "" : photoUri);
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: Probably ready to remove? :)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 6]
```diff
@@ -187,408 +245,303 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
-              break;
-            default:
-              label = "unknown";
-          }
+    // convert to react native object
+    public WritableMap toMap(Set<String> fieldSet) throws ParseException {
+      WritableMap contact = Arguments.createMap();
+      contact.putString("id", contactId);
+      contact.putString("name", !TextUtils.isEmpty(displayName) ? displayName : givenName + " " + familyName);
+      contact.putString("firstName", givenName);
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: I think if any value is null/empty we den't want to put it in the response at all. 

So, initializing fields of this class is not necessary. Perhaps a method like `put to map if not null` would do?

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 7]
```diff
@@ -61,116 +108,127 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
+      cursor = cr.query(
+          ContactsContract.Data.CONTENT_URI,
+          FULL_PROJECTION.toArray(new String[FULL_PROJECTION.size()]),
+          selection,
+          selectionArgs.toArray(new String[selectionArgs.size()]),
+          null);
+    }
+    if (cursor != null) {
+      try {
+        // populate list of contacts from cursor query
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: Some comments are probably not quite necessary - you can leave the ones that describe something tricky.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 8]
```diff
@@ -61,116 +108,127 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
+          ContactsContract.Data.CONTENT_URI,
+          FULL_PROJECTION.toArray(new String[FULL_PROJECTION.size()]),
+          ContactsContract.Data.CONTACT_ID + " = ?",
+          new String[] { options.getString("id") },
+          null
+      );
+    else {
+
+
+      String selection = ContactsContract.Data.MIMETYPE + "=? OR " + ContactsContract.Data.MIMETYPE + "=? OR " + ContactsContract.Data.MIMETYPE + "=? OR " + ContactsContract.Data.MIMETYPE + "=? OR " + ContactsContract.Data.MIMETYPE + "=?";
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: Does this work as expected? Eg. `prefix` is not added to selection but is still there in the response. If this must be there - please give this some formatting - every "or" in single line :)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 9]
```diff
@@ -61,116 +108,127 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
+      ArrayList<String> selectionArgs = new ArrayList<>(Arrays.asList(CommonDataKinds.Email.CONTENT_ITEM_TYPE, CommonDataKinds.Phone.CONTENT_ITEM_TYPE, CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE, CommonDataKinds.Organization.CONTENT_ITEM_TYPE, CommonDataKinds.StructuredPostal.CONTENT_ITEM_TYPE));
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: Similar as above : if necessary -> format - just like projection arrays are formatted

**Review by aalices — CHANGES_REQUESTED** [review]
Really sorry you had to wait so long! Quite busy times. Generally the changes look good and make `Contacts` superfast! :) 

However, I made some comments and also think we don't need any non-.java files here, am I right? :) just do not include the stuff you put to `.gitignore` to your commit.

Many thanks!

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 5 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: [X]

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 6 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: > I think if any value is null/empty we den't want to put it in the response at all.

i do agree however prioritized parity with [existing functionality](https://github.com/expo/expo/blob/master/android/expoview/src/main/java/abi26_0_0/host/exp/exponent/modules/api/ContactsModule.java#L202), which helps it's simpler for now as well :)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 7 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 8 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 9 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 4 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: overall agree on 1 class per file however between
- likely a future perf refactor to remove the pojo translation and go directly from cursor query to react native `WriteableMap`
- `Contact` class only used with `ContactsModule`
- line count of `ContactsModule` is still less than prior to changes

i thought it wasn't necessary.  Happy to move out if you still think best.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 2 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: good catch! unfortunate there isn't a good way to catch this in current test automation

check.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 6 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: did change for firstName, middleName, and lastName

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 1 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check.

good catch, definitely an un-used variable - a remnant from some testing on generisizing the "add-on fields" functionality

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 3 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: > What actually is a 'me' user?
https://android.stackexchange.com/questions/17320/how-to-edit-the-me-contact

i honestly kept it as a safety carry over from react-native-contacts (https://github.com/rt2zz/react-native-contacts/pull/55/files)

i'll remove it, shouldn't cause trouble b/c up to this wouldn't have been relevant in expo api

**Comment by tiny-dancer** [comment]
@aalices, my turn to apologize on wait time :)

updates have been made and ready to go pending confirmations on 2 feedback items

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 6 above)
Comment (aalices) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: 🤔 well, we don't put it if it is null: https://github.com/expo/expo/blob/master/android/expoview/src/main/java/abi26_0_0/host/exp/exponent/modules/api/ContactsModule.java#L221
or am I missing your point? :)

Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: hazah! you caught me :)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 1 above)
Comment (aalices) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: I also meant the other ones, are they required? Like I mentioned passing `null` seems to behave just okay but I might be wrong.

Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: image

tests start failing when i commented out all of the PROJECTION add statements. where did you pass null exactly?

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 6 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 10]
```diff
@@ -457,17 +457,23 @@ public WritableMap toMap(Set<String> fieldSet) throws ParseException {
...
+      }
       contact.putBoolean("imageAvailable", this.hasPhoto);
       if (fieldSet.contains("thumbnail")) {
         WritableMap thumbnail = Arguments.createMap();
         thumbnail.putString("uri", this.hasPhoto ? photoUri.toString() : null);
         contact.putMap("thumbnail", thumbnail);
       }
 
-      if (fieldSet.contains("note")) { // double if check with query
+      if (fieldSet.contains("note") && !TextUtils.isEmpty(department)) { // double if check with query
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: probably `note`, not `department` :)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 4 above)
Comment (aalices) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: direct translation sounds good :+1:

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 1 above)
Comment (aalices) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: https://github.com/expo/expo/pull/1349/files#diff-86247190873d660d66a885d9af8272dcR200
right here

**Comment by tiny-dancer** [comment]
> probably note, not department :)

have no idea where this comment went in the pr view, but check. good catch on my sloppiness

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 1 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: thanks for clarifying. see below quote from [android docs](https://developer.android.com/reference/android/content/ContentResolver.html#query(android.net.Uri, java.lang.String[], android.os.Bundle, android.os.CancellationSignal)):

> projection | String: A list of which columns to return. Passing null will return all columns, which is inefficient.

**Comment by aalices** [comment]
@tiny-dancer oh yes, you fixed this so fast that i thought the commit was here before I wrote the comment so I removed it 😅

**Comment by tiny-dancer** [comment]
@aalices haha
Alt Text

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 1 above)
Comment (aalices) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: okay makes sense, I got a bit lost int the docs. Anway, i think the projection still contains phone numbers, emails and addresses even if they're not in `fields`.

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 4 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: @aalices, I should have clarified `future perf refactor` would preferable be in a future pull request. are we good to proceed here for current pr?

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 1 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: @aalices updated phone numbers, emails and addresses to only be projection if requested. surprisingly had essentially no impact on performance, could look into including them by default in future

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 11]
```diff
@@ -187,408 +258,307 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
+  }
 
-          String label;
+  private static class Contact {
+    private String contactId;
+    private String displayName;
+    private String givenName = "";
+    private String middleName = "";
+    private String familyName = "";
+    private String prefix = "";
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: unused
```
if (!TextUtils.isEmpty(prefix)) {
        contact.putString("namePrefix", prefix);
      }
```
🙂

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 12]
```diff
@@ -187,408 +258,307 @@ public void getContactsAsync(final ReadableMap options, final Promise promise) {
...
+      if (!TextUtils.isEmpty(jobTitle)) {
+        contact.putString("jobTitle", jobTitle);
+      }
+      if (!TextUtils.isEmpty(department)) {
+        contact.putString("department", department);
+      }
+      contact.putBoolean("imageAvailable", this.hasPhoto);
+      if (fieldSet.contains("thumbnail")) {
+        WritableMap thumbnail = Arguments.createMap();
+        thumbnail.putString("uri", this.hasPhoto ? photoUri.toString() : null);
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: `photoUri` is a String actually :)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java [hunk 13]
```diff
@@ -42,6 +42,25 @@
 public class ContactsModule extends ReactContextBaseJavaModule {
   private static final String TAG = ContactsModule.class.getSimpleName();
 
+  private static final List<String> JUST_ME_PROJECTION = new ArrayList<String>() {{
```
Comment (aalices) [inline android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: maybe something like `BASIC_PROJECTION` would give some more idea of what it is

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 12 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check.

can never be too sure ;)

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 13 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check

File: android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java (hunk 11 above)
Comment (tiny-dancer) [reply android/expoview/src/main/java/versioned/host/exp/exponent/modules/api/ContactsModule.java]: check.

good catch, hate not being able to seed data on tests 😠

**Comment by aalices** [comment]
As we're kinda busy + planning a release soon and we want to properly test your changes, we'll get back to this PR very soon. If you think it is a good time to start coding the `future refactor`, feel free - if not - we will consider merging it as it is now. Sorry you have to wait sooo long!

**Comment by terrysahaidak** [comment]
Hello @aalices, any updates?)

**Comment by esamelson** [comment]
Hi @terrysahaidak - no updates at this time, but we promise we'll get to this PR as soon as we can. Thanks for your patience!!

**Review by aalices — APPROVED** [review]
I think we're ready to merge, just waiting for @esamelson comments :)

**Review by esamelson — APPROVED** [review]
lgtm. I'm really sorry you had to wait so long... has been a crazy few months for us here at Expo 😅 I'm going to go ahead and merge this now and it should go out with SDK 27.

**Comment by robertobrogi** [comment]
hi, sorry for the question but you have idea when is out the fix for android? because now in expo 26 the contacts in android is unusable...  and i have the app ready for the store :(

thanks

**Comment by esamelson** [comment]
Hi @robertobrogi - sorry, I can't give you any estimate for when SDK 27 will be out, but I recommend following our [blog](https://blog.expo.io/) as we will announce it there as soon as it's ready!

**Comment by terrysahaidak** [comment]
Hi @robertobrogi, unfortunately, even with this PR the app fill freeze for 1-5 seconds for getting contacts. So you need to build some workaround for this. For example, show a static screen with the spinner while fetching contacts.

If you wanna fetch the contacts in the background, there is no way for you without completely freeze the app for some time.
As a workaround for this, Expo team could provide an alternative solution for current Promise based API an event emitter based one. It won't freeze the app, and you can call it as many times as you want. For this approach, you will need an event listener. As soon as you subscribe to some `expo-contacts-fetched` event, you will get it in the background it has fetched. I know it looks like a hack or crutch, but it will work unlike the promises freeze the app completely.

**Comment by robertobrogi** [comment]
Thanks @terrysahaidak about the information. But which is the reason about this freeze? maybe the best way is take just the info you need. example if you need just name and phone number when you call getAsyncContact the return are just this two inos and in this case i think the the response are immediately. We need to specify which info you need. What you think

**Comment by terrysahaidak** [comment]
I don't know why it freezes. I haven't expected it before by calling come native promise-based API. It just blocks the JS thread and that's it.

Maybe someone from the expo team or react native team can explain why it happens.
/cc @esamelson @aalices @tiny-dancer

**Comment by tiny-dancer** [comment]
@terrysahaidak @robertobrogi apologies, i don't have legit root cause off-hand for the freezing, however will dig into it when looking into [querying capabilities](https://expo.canny.io/feature-requests/p/query-contacts).

side note: open source contributor, not expo/rn team member :)

**Comment by tiny-dancer** [comment]
@terrysahaidak @robertobrogi call me crazy, i'm surprisingly not able to reproduce the freezing - do either of you have a sample app I could look into?

**Comment by terrysahaidak** [comment]
Hi @tiny-dancer, thank you so much. I'm experience it in my almost production app, also, I'm running it on my phone there are around 600 contacts. On iOS with more than 1000 contacts, everything is good. I will make you some demo project where you will be able to experience this behavior.

**Comment by terrysahaidak** [comment]
Hi @tiny-dancer, here are examples of freezing the app (without your PR, just an Expo SDK 26, don't have an ability to test with it):

Development env: https://photos.app.goo.gl/PzZ5R5YQsrM3Ay403
Production env: https://photos.app.goo.gl/nVg3ICqRVMd4vwCo1
Repository: https://github.com/oleh-mryhlod/RN-test-expo-get-contacts

As you can see, JS Thread gets blocked when fetching contacts begin.
Could you please confirm it is fixed with your PR? Will appreciate that a lot.

**Comment by tiny-dancer** [comment]
@terrysahaidak awesome! will look to dig in in the coming days.

**Comment by tiny-dancer** [comment]
@terrysahaidak opened pull request https://github.com/expo/expo/pull/1676 to resolve freezing 
/fyi @aalices @esamelson 

as for comparison of sdk 26 vs 27

SDK 26
screenshot_expocontacts_26

SDK 27
screenshot_expocontacts_27

**Comment by terrysahaidak** [comment]
@tiny-dancer thanks a lot!

**Pull request closed without merging**